### PR TITLE
Implemented ScriptBlock support for ValidateSet

### PR DESCRIPTION
**Pull request by dkattan**
# PR Summary

Implements #8651

Adds constructor to ValidateSetAttribute that accepts a ScriptBlock. 

Adds logic that evaluates the ScriptBlock when ValidValues is accessed

Adds logic to Compiler.cs that detects the presence of a ScriptBlock and uses the new constructor

Adds test to TabCompletion.Tests.ps1 to verify functionality

## PR Context

I would like to be able to have dynamic parameter values self-contained in a function definition.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/Attributes.cs [hunk 1]
```diff
@@ -1686,11 +1686,32 @@ public IList<string> ValidValues
         {
             get
             {
-                if (validValuesGenerator == null)
+                if (validValuesGenerator == null && validValuesScript == null)
                 {
                     return _validValues;
                 }
+                if (validValuesScript is not null)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs:1693]:
```suggestion

                if (validValuesScript is not null)
```

File: src/System.Management.Automation/engine/Attributes.cs [hunk 2]
```diff
@@ -1595,7 +1595,7 @@ public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerat
         // Cached valid values.
         private string[] _validValues;
         private readonly int _validValuesCacheExpiration;
-
+
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs:1598]:
```suggestion

```

File: src/System.Management.Automation/engine/Attributes.cs [hunk 3]
```diff
@@ -1655,7 +1655,7 @@ public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
         // of 'validValuesGenerator'.
         private readonly string[] _validValues;
         private readonly IValidateSetValuesGenerator validValuesGenerator = null;
-
+        private readonly ScriptBlock validValuesScript = null;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs:1658]:
```suggestion
        private readonly ScriptBlock validValuesScript;
```

File: src/System.Management.Automation/engine/Attributes.cs [hunk 4]
```diff
@@ -1686,11 +1686,32 @@ public IList<string> ValidValues
...
+                        if (resultAsString != null)
+                        {
+                            validValuesFromScript.Add(resultAsString);
+                            continue;
+                        }
+                        var resultToString = customResult.ToString();
+                        validValuesFromScript.Add(resultToString);
+                    }
+                    return validValuesFromScript;
+                }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs:1714]:
```suggestion
                        }

                        var resultToString = customResult.ToString();
                        validValuesFromScript.Add(resultToString);
                    }

                    return validValuesFromScript;
                }

```

**Comment by dkattan** [comment]
I'm thinking about this, and it appears that we don't get access to other parameters that may be specified which would be the next logical thing you would want to have in the script block. 

@iSazonov How would you feel about providing the ScriptBlock parameters similar to that of ArgumentCompleter?

param ( $commandName,
                    $parameterName,
                    $wordToComplete,
                    $commandAst,
                    $fakeBoundParameters )

**Comment by iSazonov** [comment]
@dkattan Could you please share demo script?

**Comment by dkattan** [comment]
> @dkattan Could you please share demo script?

Inspired by the sample for ArgumentCompleters:
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_argument_completion?view=powershell-7.2#argumentcompleter-script-block

```powershell
function Test-DynamicValidateSet {
[CmdletBinding()]
 param (
        [Parameter(Mandatory=$true)]
        [ValidateSet('Fruits', 'Vegetables')]
        $Type,

        [Parameter(Mandatory=$true)]
        [ValidateSet({ 
            param ( $commandName,
            $parameterName,
            $commandAst,
            $fakeBoundParameters )

            $possibleValues = @{
                Fruits = @('Apple', 'Orange', 'Banana')
                Vegetables = @('Tomato', 'Squash', 'Corn')
            }

            if ($fakeBoundParameters.ContainsKey('Type')) {
                $possibleValues[$fakeBoundParameters.Type]
            } else {
                $possibleValues.Values | ForEach-Object {$_}
            }
        })]
        $Value
      )
}

Test-DynamicValidateSet -Type Fruits -Value 
````

**Comment by SeeminglyScience** [comment]
> I think that's why they chose to call it "fakeBoundParameters" in the ArgumentCompleter implementation. It does a pseudo-binding that seems to be best effort, which is good enough for most cases.

Yeah it's called that because binding hasn't happened yet, it's all based on statically inferred values. Best effort makes perfect sense for argument completion, but for actual binding time validation I don't think we want to surface something that inconsistent.

**Comment by dkattan** [comment]
> ```powershell
> function Test-DynamicValidateSet {
> [CmdletBinding()]
>  param (
>         [Parameter(Mandatory=$true)]
>         [ValidateSet('Fruits', 'Vegetables')]
>         $Type,
> 
>         [Parameter(Mandatory=$true)]
>         [ValidateSet({ 
>             param ( $commandName,
>             $parameterName,
>             $commandAst,
>             $fakeBoundParameters )
> 
>             $possibleValues = @{
>                 Fruits = @('Apple', 'Orange', 'Banana')
>                 Vegetables = @('Tomato', 'Squash', 'Corn')
>             }
> 
>             if ($fakeBoundParameters.ContainsKey('Type')) {
>                 $possibleValues[$fakeBoundParameters.Type]
>             } else {
>                 $possibleValues.Values | ForEach-Object {$_}
>             }
>         })]
>         $Value
>       )
> }
> 
> Test-DynamicValidateSet -Type Fruits -Value 
> ```

This is beginning to make sense. In order for this to work the way I want it to work I'd need to introduce a "DelayedValidation" feature that waits for all the parameters to be bound and then passes in the bound parameters to the ScriptBlock.

Then what happens if multiple parameters are awaiting validation? 

This is getting deep fast.

I'm going to revert the parameter change so we can at least get the basic functionality working.

If someone needs access to data in the runspace, they can always declare global variables that would be equally available to both ArgumentCompleters and ValidateSet.

**Comment by dkattan** [comment]
@SeeminglyScience  @iSazonov  I've reverted the change. How are you guys feeling about this PR?

**Comment by iSazonov** [comment]
> @SeeminglyScience @iSazonov I've reverted the change. How are you guys feeling about this PR?

It is question for WG. The PR is implementing initial proposal from #3744 but the discussion in #3744 was finished with the conclusion that we need a consistency with binary cmdlets and as result we implemented IValidateSetValuesGenerator in #3784. ( Clearly, this is not very convenient to use in scripts. We could discuss this in #8651)
So we need a review from WG-Engine before we can continue.

**Comment by dkattan** [comment]
Dumb question, what’s WG?

**Comment by dkattan** [comment]
> Dumb question, what’s WG?

Nevermind, I Googled it

https://github.com/PowerShell/PowerShell/blob/master/docs/community/working-group-definitions.md

**Comment by iSazonov** [comment]
@vexx32 Could you add this in WG plan?

**Comment by dkattan** [comment]
> @vexx32 Could you add this in WG plan?

Hey @vexx32 Any updates/eta?

**Comment by dkattan** [comment]
@vexx32 @iSazonov Any updates?

**Comment by SeeminglyScience** [comment]
The Engine-WG discussed this yesterday and we've found a few issues with this concept that will not be easily solvable.

1. Accepting a `ScriptBlock` here will imply the UX of generation based on supplied value, e.g. `$_`. I know your original draft did do something like this and it was changed based on feedback, but the implication will still be there.
2. When this attribute is applied within a module, the `ScriptBlock` invoked will not be invoked in the correct session state. This is already a problem with `ArgumentCompleterAttribute`, but we do not want to add additional features with the same problem until we have a solution.

When the latter issue is resolved, the language WG should revisit the former and determine if there is a feasible design.

**Comment by dkattan** [comment]
@SeeminglyScience I’d be happy to work on the module context issue. 

My end game is to create a UX similar to Show-Command but for the web. My project (immy.bot) is very tightly integrated with PowerShell with custom PSHosts that push Write-Progress to progress bars in the interface, for example.

We already have static forms where you can create parameters for scripts, but it gets rather clunky when the need for parameter sets arise.

Rather than reinvent the dynamic forms wheel, my thought was to simply parse the params block of the script and same parameter values as a hashtable that can splatted on to the script in question and PowerShell can handle the binding.

Obviously this can be done without the feature in the PR, however the primary use case for me does require it, I’ll elaborate.

My application installs software on computers based on the user you select. Users are dynamically retrieved from AzureAD. This is currently implemented in C#, but to offer maximum flexibility I’d like to move this logic into PowerShell so users can define their own behavior.

Now I suppose I could implement dynamic parameters, assuming that works with Show-Command/Get-Command the syntax is just very cumbersome.

I’d rather keep everything in a scriptblock and simply call Get-AzureADUser.

ANYWAY, if you wouldn’t mind, can you elaborate on the issue with module context?  Is it related to how modules imported using InitialSessionState.ImportPSModuleFromPath are not bound by constrained language mode?

**Comment by iSazonov** [comment]
I think this requires the implementation of _stateless_ script block. Theoretically it is possible, but in fact it may be difficult to implement, and most importantly lead to negative consequences for PowerShell language as a whole.

**Comment by dkattan** [comment]
> I think this requires the implementation of _stateless_ script block. Theoretically it is possible, but in fact it may be difficult to implement, and most importantly lead to negative consequences for PowerShell language as a whole.

Can you give me an example on what sort of negative consequences this would have on the language as a whole?

**Comment by iSazonov** [comment]
Today any scriptblock exists and run only in one runspace. This is how PowerShell works. 
If this principle is abandoned completely, it will be a completely different language. If partially, it could lead to chaos.

**Comment by dkattan** [comment]
It sounds like the concern is that the implementation of a stateless script block (which I’m interpreting as being a purely functional/isolated script block similar to a static method in C#) would violate the one runspace paradigm of PowerShell. I could argue that by bringing up implicit remoting for Windows PowerShell compatibility or PS Remoting in general, but I’d rather try to understand why you think a stateless scriptblock is necessary in the first place?

If not for me, for potential future readers who want to fully understand. 

And also, help me understand the existing problems with Modules

I really am trying to understand this at a deeper level, not trying to be difficult. I appreciate the time you’ve already put into this.

**Comment by iSazonov** [comment]
In remoting PowerShell doesn't send scriptblock, it send _text_. Also there is 1-t0-1 mapping local to remote runspace.
I think the main problem is how users perceive it. On the one hand they would like to make the script code independent. On the other hand it is always required to run it in some context. PowerShell historically has a solution. For example, a module has its own context. This is what the module concept exists for. It's simple and it's understandable to users. If we allow an exception, there must be an idea behind it that everyone understands. What I mean by that is that the design of the language should change.

I don't know what else to add. :-)

**Comment by SeeminglyScience** [comment]
> I think this requires the implementation of _stateless_ script block.

Sorry I think there was a misunderstanding. It just needs the right state, it doesn't need to be stateless.

@dkattan a quick way to demo what I'm talking about:

```powershell
New-Module {
    $ModuleScopedVar = 'one', 'two'

    function Test-Command {
        [CmdletBinding()]
        param(
            [ArgumentCompleter({ $script:ModuleScopedVar })]
            [string] $Name
        )
    }
} | Import-Module
```

There are workarounds but this is already a huge UX issue that comes up often on the discord. Unfortunately it wasn't caught (or maybe the impact wasn't realized) for `ArgumentCompleter`, but until it's solved we should avoid adding more things with the same behavior.

**Comment by dkattan** [comment]
@SeeminglyScience that example helped tremendously. 

To better illustrate the problem, I created a variable with the same name with value three, four and it tab completes three,four instead of one,two

It seems like the problem is that CompleteInput not taking into consideration that Test-Command is defined in a module and therefore shouldn’t have access to variables in the script scope.

If I had to create an issue, it would be “Results of CompleteInput in modules potentially affected by variables defined in parent runspace”

I understand now that if we allow ValidateSet to be dynamic and not a constant it would have the same problem.

Also, do DynamicParameters have the same problem? I’ll test shortly.

I have no issue using DynamicParameters for my use case, but I’m afraid I won’t be able to instantiate the required types in Constrained Language mode. (The scripts in my application all run in Constrained Language mode)

I suppose I could inject custom Cmdlets that instantiate these objects from C# as a workaround.

**Comment by iSazonov** [comment]
> Sorry I think there was a misunderstanding. It just needs the right state, it doesn't need to (and shouldn't) be stateless.

Yes, you mean rather "scope," I mean "sessionstate".
I suspect the problem with variable propagation like `$ErrorActionPreference` is of the same nature. This makes me think that the flexible solution could be rather related to "sessionstate" (which would certainly in turn be a means of manipulating scopes).

**Comment by SeeminglyScience** [comment]
> It seems like the problem is that CompleteInput not taking into consideration that Test-Command is defined in a module and therefore shouldn’t have access to variables in the script scope.

It's more that the `ScriptBlock` is created in a way that `ScriptBlock.SessionStateInternal` is not populated. When that is null, it implicitly takes on the state in `ExecutionContext.EngineSessionState` (whatever was last running at the time).

Curiously, `ValidateScript` does not have this problem. I'm unsure if it's explicitly special cased, or if it *happens* to work based on when the script is invoked. Either way worth looking into.

> Also, do DynamicParameters have the same problem? I’ll test shortly.

They do not! Neither does `ValidateScript` or `Register-ArgumentCompleter` (the pair of them is what I'd recommend you use to implement this feature in the mean time).

**Comment by SeeminglyScience** [comment]
> Yes, you mean rather "scope," I mean "sessionstate". I suspect the problem with variable propagation like `$ErrorActionPreference` is of the same nature. This makes me think that the flexible solution could be rather related to "sessionstate" (which would certainly in turn be a means of manipulating scopes).

`SessionState` is what I'm referring to when I say state, sorry if that was unclear.

**Pull request closed without merging**